File: python/vcp_sdk/canonical.py
"""Canonical JSON (JCS, RFC 8785) and content-addressed hashing (SPEC §3).

All content-addressing, signing, and binding in VCP depends on a single
unambiguous serialization. We canonicalize with JCS and hash with SHA-256,
emitting ``sha256:<lowercase-hex>``.

The conformance vectors (and the v0.1 wire contract) restrict values to
objects, arrays, strings, integers, booleans, and null. For that subset JCS
reduces to "sort object keys by code unit, no whitespace, UTF-8", which is
exactly what ``json.dumps(value, sort_keys=True, separators=(",", ":"),
ensure_ascii=False)`` produces. Floats are intentionally out of scope for v0.1
(their JCS number formatting is the one genuinely fiddly part).
"""

from __future__ import annotations

import hashlib
import hmac
from typing import Any

__all__ = [
    "canonical_json",
    "hash",
    "sha256_hex",
    "constant_time_equals",
]


def canonical_json(value: Any) -> bytes:
    """Return JCS (RFC 8785) canonical bytes for ``value``.

    Implemented exactly as the SDK contract requires::

        json.dumps(value, sort_keys=True, separators=(",", ":"),
                   ensure_ascii=False).encode("utf-8")
    """
    import json

    return json.dumps(
        value,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")


def sha256_hex(data: bytes) -> str:
    """Lowercase hex SHA-256 digest of raw bytes."""
    return hashlib.sha256(data).hexdigest()


def hash(value: Any) -> str:  # noqa: A001 - name mandated by the SDK contract
    """Return ``"sha256:" + hexdigest`` of ``canonical_json(value)`` (SPEC §3)."""
    return "sha256:" + sha256_hex(canonical_json(value))


def constant_time_equals(a: str, b: str) -> bool:
    """Constant-time string comparison for hash / identifier checks (SPEC §3.5).

    Identifier comparison is exact byte-for-byte; no normalization, case-folding,
    or Unicode equivalence is applied here (canonicalization happened at hash
    time). We only add timing resistance.
    """
    return hmac.compare_digest(a.encode("utf-8"), b.encode("utf-8"))

File: typescript/packages/sdk/src/canonical.ts
import { createHash } from "node:crypto";

/**
 * JSON Canonicalization Scheme (RFC 8785) for the value subset used by VCP:
 * objects, arrays, strings, integers, booleans, and null.
 *
 * Rules implemented (SPEC §3):
 *  - Object keys are sorted by UTF-16 code unit (the default lexicographic
 *    order of String comparison in JS, which is code-unit order).
 *  - No insignificant whitespace.
 *  - Strings use the minimal JSON escaping mandated by RFC 8785 §3.2.2.2,
 *    which is exactly what JSON.stringify on a single string produces: the
 *    seven short escapes (\" \\ \b \f \n \r \t), \u00xx for the remaining
 *    C0 control characters, and every other code point emitted literally as
 *    UTF-8 (no \u escaping of non-ASCII).
 *  - Integers are emitted in their shortest decimal form.
 *
 * Floating point numbers are intentionally NOT supported: RFC 8785 number
 * formatting (ECMAScript Number-to-String / shortest round-trip) is the one
 * genuinely fiddly part and is out of scope for VCP v0.1 vectors. A non-integer
 * number throws.
 */
export function canonicalJson(value: unknown): string {
  return serialize(value);
}

function serialize(value: unknown): string {
  if (value === null) return "null";

  const t = typeof value;

  if (t === "boolean") return value ? "true" : "false";

  if (t === "string") {
    // JSON.stringify of a lone string yields RFC 8785 §3.2.2.2-compliant
    // escaping: short escapes + \u00xx for control chars, literal otherwise.
    return JSON.stringify(value);
  }

  if (t === "number") {
    const n = value as number;
    if (!Number.isFinite(n)) {
      throw new TypeError(`canonicalJson: non-finite number ${n} is not representable`);
    }
    if (!Number.isInteger(n)) {
      throw new TypeError(
        `canonicalJson: non-integer number ${n} is out of scope for VCP v0.1 canonicalization`,
      );
    }
    // Integers: shortest decimal. Normalize -0 to 0.
    return Object.is(n, -0) ? "0" : String(n);
  }

  if (t === "bigint") {
    return (value as bigint).toString();
  }

  if (Array.isArray(value)) {
    return "[" + value.map((el) => serialize(el)).join(",") + "]";
  }

  if (t === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).filter((k) => obj[k] !== undefined);
    // Default Array.sort compares by UTF-16 code unit, which is what RFC 8785
    // requires for member ordering.
    keys.sort();
    const parts: string[] = [];
    for (const k of keys) {
      parts.push(JSON.stringify(k) + ":" + serialize(obj[k]));
    }
    return "{" + parts.join(",") + "}";
  }

  throw new TypeError(`canonicalJson: unsupported value type ${t}`);
}

/** SHA-256 over the UTF-8 bytes of JCS(value), as "sha256:<lowercase-hex>". */
export function hash(value: unknown): string {
  const canonical = canonicalJson(value);
  const digest = createHash("sha256").update(canonical, "utf8").digest("hex");
  return "sha256:" + digest;
}

/** Raw lowercase hex SHA-256 digest (no prefix) of arbitrary bytes/string. */
export function sha256Hex(data: string | Uint8Array): string {
  return createHash("sha256").update(data).digest("hex");
}
